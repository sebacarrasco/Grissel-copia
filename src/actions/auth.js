import jwtDecode from 'jwt-decode';
import Swal from 'sweetalert2';
import api from '../api';
import { types } from '../types/types';

export const startLogin = (email, password) => {
    return async(dispatch) => {

        try {
            const { data } = await api.post("login/user", {
                email,
                password
            });
            dispatch(login(data.token));
            const { userObject } = jwtDecode(data.token);
            Swal.fire({
                text: `Hola de nuevo ${ userObject.firstName }`,
                showConfirmButton: false,
                timer: 1500
            });
        }
        catch (e) {
            if (e.message.includes(401) || e.message.includes(404))
            {
                Swal.fire("El email y/o contrase침a son inv치lidos");
            }
            else
            {
                Swal.fire("Ha ocurrido un error, por favor vuelve a intentarlo");
            }
        }
    }
}

export const login = (token) => {
    return (dispatch) => {
        const { isAdmin } = jwtDecode(token);
        if (!isAdmin)
        {
            dispatch(userLogin(token));
        }
        else
        {
            dispatch(adminLogin(token));
        }
    }
};

export const userLogin = (token) => {
    const { id, exp, userObject } = jwtDecode(token);
    return {
        type: types.login,
        payload: {
            id,
            currentUser: userObject,
            expDate: exp,
            token
        }
    }
}

export const logout = () => ({
    type: types.logout
});

export const startAdminLogin = (email, password) => {
    return async(dispatch) => {

        try {
            const { data } = await api.post("login/admin", {
                email,
                password
            });
            dispatch(adminLogin(data.token));
            Swal.fire({
                text: `Hola de nuevo!`,
                showConfirmButton: false,
                timer: 1500
            });
        }
        catch (e) {
            if (e.message.includes(401) || e.message.includes(404))
            {
                Swal.fire("El email y/o contrase침a son inv치lidos");
            }
            else
            {
                Swal.fire("Ha ocurrido un error, por favor vuelve a intentarlo");
            }
        }
    }
}

export const adminLogin = (token) => {
    const { id, exp, isAdmin, email } = jwtDecode(token);
    return {
        type: types.login,
        payload: {
            id,
            email,
            isAdmin,
            expDate: exp,
            token
        }
    }
};