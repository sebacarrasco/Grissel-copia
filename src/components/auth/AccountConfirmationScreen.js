import React from 'react';
import { Container, Row, Button, Col } from 'react-bootstrap';


export const AccountConfirmationScreen = ({ history }) => {
    return (
        <Container
            style={{marginTop:'6rem'}}
        >
            <Row>
                <h1 className="text-white text-center">Listo! Tu cuenta ha sido creada</h1>
            </Row>
            <Row>
                <Col md={{ span: 2, offset: 5 }} xs={{span: 4, offset: 4}}>
                    <Button
                        className="btn-success w-100 mt-3"
                        onClick={ () => history.push("/auth/login") }
                    >
                        Ingresar
                    </Button>
                </Col>
            </Row>
        </Container>
    )
}
