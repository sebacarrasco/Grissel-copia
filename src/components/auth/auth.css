.screen-web {
    margin-top: 6rem;
    margin-left: 30%;
    margin-right: 30%;
}

.screen-mobile {
    margin-top: 10%
}