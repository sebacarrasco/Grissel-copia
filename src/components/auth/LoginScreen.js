import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import useMediaQuery from '../../useMediaQuery';
import { GoBackButton } from '../buttons/GoBackButton';
import validator from 'validator';
import Swal from 'sweetalert2';
import "./auth.css";
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from "react-router-dom";

export const LoginScreen = ({ startLoginAction, mode }) => {

    const { isAdmin } = useSelector(state => state.auth);

    const web = useMediaQuery("(min-width: 600px)");

    const [ formValues, handleInputChange ] = useForm({
        email: "",
        password: ""
    });
    const { email, password } = formValues;
    const dispatch = useDispatch();

    const handleSubmit = (event) => {
        event.preventDefault();
        if (!validator.isEmail(email) || password.length < 6)
        {
            Swal.fire("No se pudo iniciar sesión con los datos ingresados");
        }
        else
        {
            dispatch(startLoginAction(email, password));
        }
    }

    if (mode === "admin" && !!isAdmin)
    {
        return <Redirect to="/admin"/>
    }

    return (
        <div>
            <div className={ `text-white ${ web ? "screen-web" : "screen-mobile" }`}>
            { web ? <GoBackButton /> : null }
            <h2 className="text-center">Entra a tu cuenta { mode === "admin" ? "de administrador" : "" } </h2>

            <form onSubmit={ handleSubmit }>
                <div className="px-3">

                    <label className="mt-2 ml-2">Correo electrónico</label>
                    <br/>
                    <input
                        type="text"
                        autoComplete="off"
                        name="email"
                        value={ email }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100 mb-2"
                        style={{ borderRadius: "30px" }}
                        placeholder="jessica@gmail.com"
                    >
                    </input>
                    <br/>

                    <label className="mt-2 ml-2">Contraseña</label>
                    <br/>
                    <input
                        type="password"
                        autoComplete="off"
                        name="password"
                        value={ password }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="&#9679;&#9679;&#9679;&#9679;&#9679;&#9679;"
                    >
                    </input>
                    <br/>

                    <div className="d-flex justify-content-center">
                        <button
                            type="submit"
                            className="btn btn-success mt-4 px-5"
                            style={{ borderRadius: "30px" }}
                        >
                            Entrar
                        </button>
                    </div>

                    {
                        mode === "user"
                        ?
                        <p className="mt-5 px-3 text-center">
                            No tienes una cuenta? {" "}
                            <Link to="/auth/register" className="link-info">Crea una aquí</Link>
                        </p>
                        :
                        null
                    }

                </div>
            </form>
        </div>
        </div>
    )
}
