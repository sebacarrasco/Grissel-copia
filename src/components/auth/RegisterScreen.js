import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import useMediaQuery from '../../useMediaQuery';
import { GoBackButton } from '../buttons/GoBackButton';
import validator from 'validator';
import Swal from 'sweetalert2';
import "./auth.css";
import { rutTools } from 'prettyutils';
import API from '../../api';

export const RegisterScreen = ({ history }) => {

    const web = useMediaQuery("(min-width: 600px)");

    const [ formValues, handleInputChange ] = useForm({
        firstName: "",
        lastName: "",
        rut: "",
        email: "",
        password: "",
        passwordConfirmation: ""
    });

    const {
        firstName,
        lastName,
        rut,
        email,
        password,
        passwordConfirmation
    } = formValues;

    const handleSubmit = (event) => {
        event.preventDefault();
        if (!validator.isEmail(email))
        {
            Swal.fire("El email ingresado no es válido");
        }
        else if (firstName.length < 2)
        {
            Swal.fire("El nombre ingresado no es válido");
        }
        else if (lastName.length < 2)
        {
            Swal.fire("El apellido ingresado no es válido");
        }
        else if (!rutTools.validate(rut))
        {
            Swal.fire("El rut ingresado no es válido");
        }
        else if (password.length < 6)
        {
            Swal.fire("La contraseña debe ser de al menos 6 caracteres");
        }
        else if (password !== passwordConfirmation)
        {
            Swal.fire("Las contraseñas deben ser iguales");
        }
        else
        {
            API.post("users", formValues)
            .then(response => {
                Swal.fire("Solo falta un paso", "Debes hacer click en el link del email que te acabamos de enviar");
                // history.push("/login");
            })
            .catch(e => {
                if (e.message.includes(400))
                {
                    Swal.fire("Ya existe una cuenta asociada a ese correo electrónico o rut");
                }
                else
                {
                    Swal.fire("Ha ocurrido un error, por favor vuelve a intentarlo");
                }
            });
        }
    }

    return (
        <div>
            <div className={ `text-white ${ web ? "screen-web" : "screen-mobile" }`}>
            { web ? <GoBackButton /> : null }
            <h2 className="text-center">Crea tu cuenta</h2>

            <form onSubmit={ handleSubmit }>
                <div className="px-3">

                <label className="mt-2 ml-2">Nombre</label>
                    <br/>
                    <input
                        type="text"
                        autoComplete="off"
                        name="firstName"
                        value={ firstName }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="Jessica"
                    >
                    </input>
                    <br/>

                    <label className="mt-2 ml-2">Apellido</label>
                    <br/>
                    <input
                        type="text"
                        autoComplete="off"
                        name="lastName"
                        value={ lastName }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="Graham"
                    >
                    </input>
                    <br/>

                    <label className="mt-2 ml-2">Rut</label>
                    <br/>
                    <input
                        type="text"
                        autoComplete="off"
                        name="rut"
                        value={ rut }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="21.200.425-5"
                    >
                    </input>
                    <br/>

                    <label className="mt-2 ml-2">Email</label>
                    <br/>
                    <input
                        type="text"
                        autoComplete="off"
                        name="email"
                        value={ email }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="jessica@gmail.com"
                    >
                    </input>
                    <br/>

                    <label className="mt-2 ml-2">Contraseña</label>
                    <br/>
                    <input
                        type="password"
                        autoComplete="off"
                        name="password"
                        value={ password }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="&#9679;&#9679;&#9679;&#9679;&#9679;&#9679;"
                    >
                    </input>
                    <br/>

                    <label className="mt-2 ml-2">Contraseña</label>
                    <br/>
                    <input
                        type="password"
                        autoComplete="off"
                        name="passwordConfirmation"
                        value={ passwordConfirmation }
                        onChange={ handleInputChange }
                        required="required"
                        className="w-100"
                        style={{ borderRadius: "30px" }}
                        placeholder="&#9679;&#9679;&#9679;&#9679;&#9679;&#9679;"
                    >
                    </input>
                    <br/>

                    <div className="d-flex justify-content-center">
                        <button
                            type="submit"
                            className="btn btn-success mt-4 px-5"
                            style={{ borderRadius: "30px" }}
                        >
                            Crear
                        </button>
                    </div>

                    <p className="mt-3 px-3 text-center">
                        Ya tienes una cuenta? {" "}
                        <Link to="/auth/login" className="link-info">Entra aquí</Link>
                    </p>

                </div>
            </form>
        </div>
        </div>
    )
}
