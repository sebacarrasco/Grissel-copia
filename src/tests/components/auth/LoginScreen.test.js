import React from 'react';
import { mount } from 'enzyme';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';

// Config de la store
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {}
};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <LoginScreen />
        </MemoryRouter>
    </Provider>
);

describe('Tests for >LoginScreen />', () => {
   
    beforeEach( () => {
        store = mockStore(initState);
        jest.clearAllMocks();
    });

    test('should match the snapshot', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

});
