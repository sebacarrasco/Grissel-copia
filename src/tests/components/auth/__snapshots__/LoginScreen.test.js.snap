// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for >LoginScreen /> should match the snapshot 1`] = `
<div>
  <h1>
    Login screen
  </h1>
</div>
`;
